src/Generator/TypeGenerator.php

```diff
@@ -49,88 +51,160 @@ final class TypeGenerator implements GeneratorInterface
+            )
+        ));
+    }
+
+    /** @psalm-pure */
+    private static function reflectionNamedTypeToString(
+        ReflectionNamedType $type,
+        ?ReflectionClass $currentClass
+    ): string {
+        $lowerCaseName = strtolower($type->getName());
 
-        if ($wasTrimmed && $isInternalPhpType) {
-            throw new InvalidArgumentException(sprintf(
-                'Provided type "%s" is an internal PHP type, but was provided with a namespace separator prefix',
-                $type
-            ));
+        if ('mixed' === $lowerCaseName || 'null' === $lowerCaseName) {
+            // `mixed` and `null` are implicitly nullable, therefore we need to skip adding nullability markers to it
+            return $lowerCaseName;
         }
 
-        if ($nullable && $isInternalPhpType && 'void' === strtolower($trimmedType)) {
-            throw new InvalidArgumentException(sprintf('Provided type "%s" cannot be nullable', $type));
+        $nullabilityMarker = $type->allowsNull()
+            ? '?'
+            : '';
+
+        if ('self' === $lowerCaseName && $currentClass) {
+            return $nullabilityMarker . $currentClass->getName();
         }
 
-        $instance = new self();
+        if ('parent' === $lowerCaseName && $currentClass && $parentClass = $currentClass->getParentClass()) {
+            return $nullabilityMarker . $parentClass->getName();
+        }
 
-        $instance->type              = $trimmedType;
-        $instance->nullable          = $nullable;
-        $instance->isInternalPhpType = $isInternalPhpType;
+        return $nullabilityMarker . $type->getName();
+    }
 
-        return $instance;
+    /**
+     * @throws InvalidArgumentException
+     *
+     * @psalm-pure
+     */
+    public static function fromTypeString(string $type): self
+    {
+        [$nullable, $trimmedNullable] = self::trimNullable($type);
+        $types = array_map([AtomicType::class, 'fromString'], explode('|', $trimmedNullable));
+
+        usort(
+            $types,
+            static fn(AtomicType $left, AtomicType $right): int => [$left->sortIndex,
+                                                                    $left->type] <=> [$right->sortIndex, $right->type]
+        );
+
+        if (1 === count($types)) {
+            $types[0]->assertCanBeAStandaloneType();
+
+            if ($nullable) {
+                $types[0]->assertCanBeStandaloneNullable();
+            }
+        } else {
+            if ($nullable) {
+                throw new InvalidArgumentException(sprintf(
+                    'Type "%s" is a union type, and therefore cannot be also marked nullable with the "?" prefix',
+                    $type
+                ));
+            }
+
+            foreach ($types as $index => $atomicType) {
+                $atomicType->assertCanUnionWith(array_diff_key($types, array_flip([$index])));
+            }
+        }
+
+        return new self($types, $nullable);
```

Comment:
```suggestion
        if (1 === count($types)) {
            $types[0]->assertCanBeAStandaloneType();

            if ($nullable) {
                $types[0]->assertCanBeStandaloneNullable();
            }
            return new self($types, $nullable);
        }
        
        if ($nullable) {
            throw new InvalidArgumentException(sprintf(
                'Type "%s" is a union type, and therefore cannot be also marked nullable with the "?" prefix',
                $type
            ));
        }
        
        foreach ($types as $index => $atomicType) {
            $atomicType->assertCanUnionWith(array_diff_key($types, array_flip([$index])));
        }

        return new self($types, $nullable);
```

I know my suggestion adds a redundant line, but it makes understanding the paths through the method a bit clearer.

Reply:
Removing one nesting level sounds fine - need to try it out and see how it looks :-)

Reply:
fc7d129f5d97cbb34f6b516a19ab5ac2d36f1d66